src/lib.rs

```diff
@@ -0,0 +1,555 @@
+                let cur = self.edges.get(i).unwrap();
+                if !visited.contains(&cur.target) {
+                    print!("{}->", cur.target);
+                    visited.insert(cur.target.clone());
+                    stack.push_front(cur.target.clone());
+                    break;
+                }
+                i += 1;
+            }
+            if i == neighbour_end.clone() {
+                stack.pop_front();
+            }
+        }
+    }
+}
+
+/// Assistance function for initial edge set
+fn update_edge_set<T>(
+    edges: &mut HashMap<T, Vec<Edge<T>>>,
+    from: T,
+    target: T,
+    attrs: HashMap<String, Attr>,
+) where
+    T: Eq + Hash + Clone,
+{
+    match edges.get(&from) {
+        Some(_v) => {
+            let mut edge_vec = edges.remove(&from).unwrap();
+            edge_vec.push(Edge {
+                source: from.clone(),
+                target,
+                attrs,
+            });
+            edges.insert(from.clone(), edge_vec);
+        }
+        None => {
+            let mut vec = Vec::new();
+            vec.push(Edge {
+                source: from.clone(),
+                target,
+                attrs,
+            });
+            edges.insert(from.clone(), vec);
+        }
+    }
+}
+
+/// Structure `Bfs` used to bind a graph and visit nodes of the bind graph in a breadth-first-search (BFScc).
+///
+/// The Structure implements trait `Iterator`, so you can use a iterator style for accessing node.
+///
+/// This algorithm binding a reference of the first given param graph which is the target traversal graph.
+///
+/// This algorithm start at the second given param `start` and traverses nodes by BFS.
+///
+/// Traversal status are storing in structure `Bfs`.
+///
+/// # Example
+///
+/// ```
+///  use graph_rust::{Bfs, Graph, GraphType};
+///  let graph = Graph::default()
+///        .with_dir(GraphType::Directed)
+///        .with_schema(String::from("..."))
+///        .from_file(String::from("data_edges.csv"), None, ",")
+///        .build();
+///  let mut bfs = Bfs::new(&graph, 1);
+///  while let Some(node) = bfs.next() {
+///        print!("{}->", node);
+///  }
+/// ```
+pub struct Bfs<'a, T> {
+    pub bind_graph: &'a Graph<T>,
+    pub stack: VecDeque<T>,
+    pub visited: HashSet<T>,
+}
+
+impl<'a, T> Bfs<'a, T>
+    where
+        T: Copy + PartialEq + Eq + Hash,
+{
+    pub fn new(graph: &'a Graph<T>, start: T) -> Self {
+        let mut bfs = Bfs {
+            bind_graph: graph,
+            stack: VecDeque::new(),
+            visited: HashSet::new(),
+        };
+        bfs.visited.insert(start);
+        bfs.stack.clear();
+        bfs.stack.push_front(start);
+        bfs
+    }
+}
+
+impl<'a, T> Iterator for Bfs<'a, T>
+    where
+        T: Copy + PartialEq + Eq + Hash,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let top_node_opt = self.stack.pop_front();
+        if let None = top_node_opt {
```

Comment:
```
if let Some(node) = top_node_opt {
    ...
} else {
    None
}
```

Reply:
If we use the template here, the code in `Some(node)` block need to move right for a tab (in rust here two space). While if there is a lot of  `if block` with a larger code size, it will cause the code likes a slide. And I think this modification will not bring some improvement in performance. So, can this part of the code remain the origin style? Or you have some other suggestions.

Reply:
Can you give me an example of what you are saying? Tell me why it would be better to stick to the original style.

Reply:
```
fn fun_with_lots_of_if_block(){
    if let Some(v1) = node_opt1{
        if let Some(v2) = node_opt2{
            if let Some(v3) = node_opt3{
                ...program the main code here
            }else{
                None
            }
        }else{
            None
        }
    }else{
        None
    }
}
```
```
fn fun_estimate_if_blocks(){
    if node_opt1.is_none(){
        return None;
    }
    if node_opt2.is_none(){
        return None;
    }
    if node_opt3.is_none(){
        return None;
    }
    ...program the main code here
}
```
Just as the two examples showing above. The second one is more flat than the first one.

Reply:
If this is the case, yes, your way will be better. What I mean is in your original case, typically it is the following case:
```
if opt.is_none() {
   return ..
}

let val = opt.unwrap();
// use val to do something.
...
```

Then using this would be better
```
if let Some(val) = opt {
   // do something with val.
} else {
    None
}
```

I am not saying you have to stick to this way of expression, but always choose whatever that is more elegant.

Reply:
I got it, thank you for your explanation.